apl-core/src/main/java/com/apollocurrency/aplwallet/apl/eth/contracts/dex.sol

```diff
@@ -0,0 +1,875 @@
+
+        IERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+        _deposit(orderId, msg.sender, token, amount);
+    }
+
+    function withdraw(
+        uint256 orderId
+    ) external {
+        // validation of the user existion
+        require(_doesUserExist(msg.sender) == true, "withdraw: the user is not active");
+
+        // storing the order information (asset and amount)
+        OrderDetails memory order = _getDepositedOrderDetails(orderId, msg.sender);
+        address asset = order.asset;
+        uint256 amount = order.amount;
+
+        // order amount validation, it should not be zero
+        require(order.withdrawn == false, "withdraw: this order Id has been already withdrawn or waiting for the swap");
+
+        _withdrawOrderBalance(orderId, msg.sender);
+
+        if (asset == _ethAssetIdentificator) {
+            msg.sender.transfer(amount);
+        } else {
+            IERC20(asset).safeTransfer(msg.sender, amount);
+        }
+
+        emit AssetWithdrawal(orderId, msg.sender, asset, amount);
+    }
+
+    // -----------------------------------------
+    // INTERNAL
+    // -----------------------------------------
+
+    function _deposit(
+        uint256 orderId,
+        address sender,
+        address asset,
+        uint256 amount
+    ) internal {
+        _activateIfUserIsNew(sender);
+        _depositOrderBalance(orderId, sender, asset, amount);
+
+        _users[sender].index += 1;
+
+        emit AssetDeposited(orderId, sender, asset, amount);
+    }
+
+    function _doesUserExist(
+        address user
+    ) internal view returns (bool) {
+        return _users[user].exist;
+    }
+
+    function _activateIfUserIsNew(
+        address user
+    ) internal returns (bool) {
+        if (_doesUserExist(user) == false) {
+            _users[user].exist = true;
+        }
+        return true;
+    }
+
+    function _getDepositedOrderDetails(
+        uint256 orderId,
+        address user
+    ) internal view returns (OrderDetails memory order) {
+        return _users[user].orders[orderId];
+    }
+
+    function _depositOrderBalance(
+        uint256 orderId,
+        address user,
+        address asset,
+        uint256 amount
+    ) internal returns (bool) {
+        _users[user].orderIdByIndex[_users[user].index] = orderId;
+        _users[user].orders[orderId] = OrderDetails(true, asset, amount, false, block.timestamp);
+        return true;
+    }
+
+    function _withdrawOrderBalance(
+        uint256 orderId,
+        address user
+    ) internal returns (bool) {
+        _users[user].orders[orderId].withdrawn = true;
+        return true;
+    }
+
+    // -----------------------------------------
+    // GETTERS
+    // -----------------------------------------
+
+    function doesUserExist(
+        address user
+    ) external view returns (bool) {
+        return _doesUserExist(user);
+    }
+
+    function getUserFilledDeposits(
```

Comment:
Looks like already withdrawn orders will be also returned

Reply:
yes, you are right.